docs/DifferentiableProgrammingImplementation.md

```diff
@@ -0,0 +1,1005 @@
+
+### Syntax
+
+The `@differentiable` declaration attribute marks "function-like" declarations ([function declarations](https://docs.swift.org/swift-book/LanguageGuide/Functions.html#ID159), [initializers](https://docs.swift.org/swift-book/LanguageGuide/Initialization.html), [properties](https://docs.swift.org/swift-book/LanguageGuide/Properties.html#ID608), and [subscripts](https://docs.swift.org/swift-book/LanguageGuide/Subscripts.html)) as differentiable.
+
+> TODO: Add `@differentiable` declaration attribute examples.
+
+The `@differentiable` declaration attribute has a few components:
+- Parameter indices clause (optional): `wrt: x, y`.
+    - This explicitly states that the declaration is to be differentiated with respect to specific parameters.
+    - If the parameter indices clause is omitted, then the parameters for differentiation are inferred to be all parameters that conform to `Differentiable`.
+- Generic where clause (optional): `where T: Differentiable`.
+    - This states that the declaration is differentiable only when certain additional generic requirements are met.
+
+### Type checking
+
+`@differentiable` declaration attributes are type-checked:
+- All `wrt` parameters of the declaring function must conform to `Differentiable` and must be specified in ascending order. If `wrt` parameters are inferred, there must be at least one `wrt` parameter.
+- The result of the declaring function must conform to `Differentiable`.
+- If JVP/VJP function names are specified, they must resolve to function declarations with the expected JVP/VJP function types (computed from the original function type, parameter indices, and generic where clause).
+
+### SIL generation
+
+`@differentiable` declaration attributes are lowered to SIL differentiability witnesses. SIL differentiability witnesses have similar components to `@differentiable` attributes, including parameter indices, result indices (a SIL implementation detail), JVP/VJP SIL function names, and an optional generic where clause.
+
+### Differentiation transform
+
+The differentiation transform canonicalizes SIL differentiability witnesses: all SIL differentiability witnesses are filled in with JVP/VJP functions.
+
+```swift
+// Before: empty differentiability witness, lowered from a `@differentiable`
+// declaration attribute.
+
+// differentiability witness for foo(_:)
+sil_differentiability_witness hidden [parameters 0] [results 0] @$s12diff_witness3fooyS2fF : $@convention(thin) (Float) -> Float {
+}
+
+// After: canonicalized differentiability witness. The differentiation transform
+// fills in JVP/VJP functions.
+
+sil_differentiability_witness hidden [parameters 0] [results 0] @$s3fooAAyS2fF : $@convention(thin) (Float) -> Float {
+  jvp: @AD__$s3fooAAyS2fF__jvp_src_0_wrt_0 : $@convention(thin) (Float) -> (Float, @owned @callee_guaranteed (Float) -> Float)
+  vjp: @AD__$s3fooAAyS2fF__vjp_src_0_wrt_0 : $@convention(thin) (Float) -> (Float, @owned @callee_guaranteed (Float) -> Float)
```

Comment:
Note that these symbols are no longer mangled this way.  We should start a separate patch and update these function names. You can write a simple program with `foo(_:)`, compile with `swiftc -emit-sil` and find its JVP and VJP's mangled names in the corresponding differentiability witness.

Reply:
I will leave that up to you, as I am not that familiar with the compiler yet.

Reply:
@rxwei Also, I need to import the Differentiation package to make the code compile. This means I can only compile it in Xcode, not in the command line (without more knowledge of the compiler). If you have a development toolchain installed, then you might have better luck with it.

Reply:
`_Differentiation` is always included in development toolchains. You just need to import `_Differentiation`. That works with command line as well.

Reply:
@rxwei awesome!  The decoration scheme was something that really bothered me.  Is the current scheme documented?

Reply:
Everything's documented in https://github.com/apple/swift/blob/main/docs/ABI/Mangling.rst (search for "autodiff" and "differentiable").

Reply:
<img width="1257" alt="Screen Shot 2021-12-07 at 8 06 00 PM" src="https://user-images.githubusercontent.com/71743241/145130139-f36af3e6-5680-4648-a59d-d9de3bd67c0c.png">

I don't have a development toolchain configured. If it's not too much to ask, could @compnerd or @rxwei fix the names?

Reply:
You don't need to fix the names to merge this PR.  We can do it later.